text/0000-gen-fn.md

```diff
@@ -0,0 +1,429 @@
+and still get a lazy iterator of values.
+
+As an example, here are multiple ways to write an iterator over something that contains integers,
+only keep the odd integers, and multiply all of them by 2:
+
+```rust
+// `Iterator` methods
+fn odd_dup(values: impl Iterator<Item = u32>) -> impl Iterator<Item = u32> {
+    values.filter(|value| value.is_odd()).map(|value| value * 2)
+}
+
+// `struct` and manual `impl`
+fn odd_dup(values: impl Iterator<Item = u32>) -> impl Iterator<Item = u32> {
+    struct Foo<T>(T);
+    impl<T: Iterator<Item = u32>> Iterator<Item = u32> for Foo<T> {
+        type Item = u32;
+        fn next(&mut self) -> Option<u32> {
+            loop {
+                let value = self.0.next()?;
+                if value.is_odd() {
+                    return Some(x * 2)
+                }
+            }
+        }
+    }
+    Foo(values)
+}
+
+// `gen block`
+fn odd_dup(values: impl Iterator<Item = u32>) -> impl Iterator<Item = u32> {
+    #[rustc_gen] {
+        for value in values {
+            if value.is_odd() {
+                yield value * 2;
+            }
+        }
+    }
+}
+
+// `gen fn`
+#[rustc_gen]
+fn odd_dup(values: impl Iterator<Item = u32>) -> u32 {
+    for value in values {
+        if value.is_odd() {
+            yield value * 2;
+        }
+    }
+}
+```
+
+Iterators created with `#[rustc_gen]` return `None` once they `return` (implicitly at the end of the scope or explicitly with `return`).
+See [the unresolved questions][#unresolved-questions] for whether `#[rustc_gen]` iterators are fused or may behave strangely after having returned `None` once.
+Under no circumstances will it be undefined behavior if `next` is invoked again after having gotten a `None`.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## New keyword
+
+Starting in the 2024 edition, `gen` is a keyword that cannot be used for naming any items or bindings. This means during the migration to the 2024 edition, all variables, functions, modules, types, ... named `gen` must be renamed.
+
+## Returning/finishing an iterator
+
+`#[rustc_gen]` block's trailing expression must be of the unit type or the block must diverge before reaching its end.
+
+### Diverging iterators
+
+For example, a `#[rustc_gen]` block that produces the infinite sequence `0, 1, 0, 1, 0, 1, ...`, will never return `None`
+from `next`, and only drop its captured data when the iterator is dropped.
+
+```rust
+#[rustc_gen] {
+    loop {
+        yield 0;
+        yield 1;
+    }
+}
+```
+
+If a `#[rustc_gen]` block panics, the behavior is very similar to `return`, except that `next` unwinds instead of returning `None`.
+
+## Error handling
+
+Within `#[rustc_gen]` blocks, the `?` operator desugars differently from how it desugars outside of `#[rustc_gen]` blocks.
+Instead of returning the `Err` variant, `foo?` yields the `Err` variant and then `return`s immediately afterwards.
```

Comment:
```suggestion
If `foo` evaluates to the `Err` variant, then instead of returning it, the desugaring of `foo?` now yields the `Err` variant and then `return`s immediately afterwards.
```

(I'm torn about this suggestion, to be honest. My motivation for suggesting it is that I thought it was not 100% clear that you were strictly discussing the case where `foo` itself returns `Err` here.)

Also: I assume the interaction with `From::from` is the same as it is for `?` outside of `#[rustc_gen]` blocks?

Reply:
(Oh, maybe this is all covered in the Reference-level explanation...)

Reply:
We've now revised this section to be more clear and comprehensive.  Thanks for pointing out the ambiguities here.